cilium/CFP-22215-mutual-auth-for-service-mesh.md

```diff
@@ -0,0 +1,442 @@
+
+#### Chosen alternative: N-tuple match on connection endpoints in new auth map
+
+A new map is introduced into the eBPF dataplane that allows the packet-handling logic to look up for some subset of `{src, dst, …}` fields of the packet and confirm whether authentication has already occurred for this set of peers.
+Pros
+Allow multiple connections between the same peers to leverage the same authentication channel, rather than forcing every connection to individually authenticate
+Could support pre-authentication + pre-population (subject to scale concerns)
+Cons
+More expensive per-packet to look up an additional map using packet fields
+Additional memory overhead (~N MB)
+
+Implementation details
+Q: How big to make the auth map in BPF? Depends on the key type and target number of peers in the cluster. Assuming ~16B for IPv4 + ~40B for IPv6, => 56 (round to 64B), 64K peers could be supported in ~4MB of memory. This comes down to 1.5MB or less if keyed by identity.
+Key
+We will only support 1:1 certificate to identity, the key could be the local+remote identities:
+```c++
+struct auth_key {
+	__u32 local_id;
+	__u32 remote_id;
+}
+```
+Alternatively, we may need to also key by destination node (since that's what we need to know to do an authentication check, and we can't assume that _all_ nodes can accept traffic for an identity)
+
+#### Alternative considered: Authentication confirmed bit in connection tracker
+* The connection tracker gains a new bit in the ct_entry field that confirms this connection is authenticated. Each connection must be independently authenticated by the control plane.
+  * Pros
+    * Simple
+    * Inherently tied to the connection that is authenticated
+  * Cons
+    * First packet of every connection that requires authentication is always dropped.
+    * Requires at least one syscall per connection for userspace to confirm authentication.
+    * Userspace CT map updates change access pattern: Could introduce jitter for all connections, regardless of whether auth is enabled or not.
+
+
+## Future Milestones
+List things that this CFP will enable but that are out of scope for now. This can help understand the greater impact of a proposal without requiring to extend the scope of a CFP unnecessarily.
+
+The following milestones are not necessarily phrased in dependency ordering, some may be doable in parallel in future.
+### Deferred Milestone 1
+Allow Cilium's Envoy Ingress/Gateway to participate in the mutual authentication mesh
+
+### Deferred Milestone 2
+Validate that a workload is supposed to be running on the target node. Mitigate cert leaks where adversary launches pods on a different node.
+
+### Deferred Milestone 3
+Negotiate encryption key in mTLS connection
+
+### Deferred Milestone 4
+Transmit metadata on authentication connection. What is the identity of the sender? Does not require the application to be running HTTP, encoding in HTTP headers in-band
```

Comment:
Which metadata do we want? A lot is already in the X.509 peer certificate

Reply:
tbh I'm not sure, Joe put this one in and I left it there. I've removed, as it's confusing.